Mathlib/Combinatorics/Additive/FreimanHom.lean

```diff
@@ -0,0 +1,369 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Algebra.BigOperators.Multiset.Basic
+import Mathlib.Data.FunLike.Basic
+import Mathlib.Data.Set.Pointwise.Basic
+
+#align_import algebra.hom.freiman from "leanprover-community/mathlib"@"f694c7dead66f5d4c80f446c796a5aad14707f0e"
+
+/-!
+# Freiman homomorphisms
+
+In this file, we define Freiman homomorphisms and isomorphism.
+
+An `n`-Freiman homomorphism from `A` to `B` is a function `f : α → β` such that `f '' A ⊆ B` and
+`f x₁ * ... * f xₙ = f y₁ * ... * f yₙ` for all `x₁, ..., xₙ, y₁, ..., yₙ ∈ A` such that
+`x₁ * ... * xₙ = y₁ * ... * yₙ`. In particular, any `MulHom` is a Freiman homomorphism.
+
+An `n`-Freiman isomorphism from `A` to `B` is a function `f : α → β` bijective between `A` and `B`
+such that `f x₁ * ... * f xₙ = f y₁ * ... * f yₙ ↔ x₁ * ... * xₙ = y₁ * ... * yₙ` for all
+`x₁, ..., xₙ, y₁, ..., yₙ ∈ A`. In particular, any `MulEquiv` is a Freiman isomorphism.
+
+They are of interest in additive combinatorics.
+
+## Main declaration
+
+* `IsMulFreimanHom`: Predicate for a function to be a multiplicative Freiman homomorphism.
+* `IsAddFreimanHom`: Predicate for a function to be an additive Freiman homomorphism.
+* `IsMulFreimanIso`: Predicate for a function to be a multiplicative Freiman isomorphism.
+* `IsAddFreimanIso`: Predicate for a function to be an additive Freiman isomorphism.
+
+## Implementation notes
+
+In the context of combinatorics, we are interested in Freiman homomorphisms over sets which are not
+necessarily closed under addition/multiplication. This means we must parametrize them with a set in
+an `AddMonoid`/`Monoid` instead of the `AddMonoid`/`Monoid` itself.
+
+## References
+
+[Yufei Zhao, *18.225: Graph Theory and Additive Combinatorics*](https://yufeizhao.com/gtac/)
+
+## TODO
+
+* `MonoidHomClass.isMulFreimanHom` could be relaxed to `MulHom.toFreimanHom` by proving
+  `(s.map f).prod = (t.map f).prod` directly by induction instead of going through `f s.prod`.
+* Affine maps are Freiman homs.
+-/
+
+open Multiset Set
+open scoped Pointwise
+
+variable {F α β γ : Type*}
+
+section CommMonoid
+variable [CommMonoid α] [CommMonoid β] [CommMonoid γ] {A A₁ A₂ : Set α}
+  {B B₁ B₂ : Set β} {C : Set γ} {f f₁ f₂ : α → β} {g : β → γ} {m n : ℕ}
+
+/-- An additive `n`-Freiman homomorphism from a set `A` to a set `B` is a map which preserves sums
+of `n` elements. -/
+structure IsAddFreimanHom [AddCommMonoid α] [AddCommMonoid β] (n : ℕ) (A : Set α) (B : Set β)
+    (f : α → β) : Prop where
+  mapsTo : MapsTo f A B
```

Comment:
In this story, the role of `B` is different from the role of `A`. In fact, one could imagine stating `IsAddFreimanHom n A` (without any reference to `B`) and adding `MapsTo f A B` when needed. This would be slightly different from what is done in paper math but maybe more natural here.

Reply:
This is how it was done previously (see `Algebra.Group.Freiman`) and it was painful.

Mathematically, removing `B` is wrong since, as explained above, `A`, `B` are the spaces defined extrinsically through ambient groups `α`, `β`.

Pragmatically, removing `B` is annoying because it makes the composition of Freiman homs not necessarily be a Freiman hom.

Reply:
The second argument looks legit to me.